_posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md

```diff
@@ -0,0 +1,54 @@
+---
+layout: post
+title: "How Kani helped find bugs in Hifitime"
+---
+
+Hifitime is a scientifically accurate time management library that provides nanosecond precision of durations and time computations for 65 thousand years, just as suitable for desktop applications as for embedded processors without a floating point unit (FPU). The purpose of this blog post is to show how Kani helped solve a number of important non-trivial bugs in hifitime.
+
+For completeness' sake, let's start with [an introduction of the notion of time](#why-time-matters), how hifitime [avoids loss of precision](#avoiding-loss-of-precision), and finally [why Kani is crucial](#importance-of-kani-in-hifitime) to ensuring the correctness of hifitime.
+
+## Why time matters
+
+_Time is just counting subdivisions of the second, it's a trivial and solved problem._ Well, it's not quite trivial, and often incorrectly solved.
+
+As explained in "Beyond Measure" by James Vincent, measuring the passing of time has been important for societies for millennia. From using the flow of the Nile River to measure the passage of a year, to tracking the position of the Sun in the sky to delimit the start and end of a day, the definition of the time has changed quite a bit. 
+
+On Earth, humans follow a local time, which is supposed to be relatively close to the idea that noon is when the Sun is at its height in the sky for the day. We've also decided that the divisions of time must be fixed: for example, an hour lasts 3600 seconds regardless of whether it's 01:00 or 14:00. If we only follow this fixed definition of an hour, then time drifts because the Earth does not in fact complete a full rotation on itself in _exactly_ 23 hours 56 minutes and 4.091 seconds (that's the duration of a sidereal day). For "human time" (Universal Coordinated Time, `UTC`) to catch up with the time with respect to the stars (`UT1`), we regularly introduce "leap seconds." Like leap years, where we introduce an extra day every fourth year (with rare exceptions), leap seconds introduce an extra seconds _every so often_, as [announced by the IERS](https://www.ietf.org/timezones/data/leap-seconds.list). When scientists report to the IERS that the Earth is about to drift quite a bit compared to `UTC`, IERS announces that on a given day at least six months in the future, there will be an extra second in that day to allow for the rotation of the Earth to catch up. In practice, this typically means that UTC clocks "stop" the time for one second: this means UTC is _not_ a continuous time scale: that second we didn't count in UTC still happened for the universe! As the Standards Of Fundamental Astronomy (SOFA) eloquently puts it:
```

Comment:
s/Earths does not in fact complete a full rotation/Earth does in fact complete a full rotation/

Reply:
That's the point ;-) The 23 hours, 56 minutes and 4.091 seconds is an average duration of a sidereal day. In practice, this changes every day. In fact, that's what the UT1 time scale measures: the _exact_ duration it took for the Earth to perform a full rotation with respect to the stars.

Reply:
Clarified in the revision, let me know if it still needs work